src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -452,7 +466,24 @@ public async Task ReadResponseHeadersAsync()
                 (wait, emptyResponse) = TryEnsureHeaders();
                 if (wait)
                 {
-                    await GetWaiterTask().ConfigureAwait(false);
+                    using (cancellationToken.Register(s => {
+                            bool signalWaiter;
+                            Http2Stream stream = (Http2Stream)s;
+                            lock (stream.SyncObject)
+                            {
+                                signalWaiter = stream._hasWaiter;
+                                stream._hasWaiter = false;
+                            }
+                            if (signalWaiter) stream._waitSource.SetResult(false);
```

Comment:
I'm not 100% sure this is safe.  It's been a while since I looked at the code for this but will want to spend some time staring at all the relevant uses tomorrow.

Reply:
It would be great if you can take another  look @stephentoub. This mimics pattern used in other places where we check if somebody is waiting and set it to false under lock and that we signal waiter outside of the lock.